// ali-rds/lib/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RDSClient = void 0;
const node_util_1 = require("node:util");
const mysql_1 = __importDefault(require("mysql"));
const operator_1 = require("./operator");
const connection_1 = require("./connection");
const transaction_1 = require("./transaction");
const literals_1 = __importDefault(require("./literals"));
class RDSClient extends operator_1.Operator {
    static get literals() { return literals_1.default; }
    static get escape() { return mysql_1.default.escape; }
    static get escapeId() { return mysql_1.default.escapeId; }
    static get format() { return mysql_1.default.format; }
    static get raw() { return mysql_1.default.raw; }
    #pool;
    constructor(options) {
        super();
        this.#pool = mysql_1.default.createPool(options);
        [
            'query',
            'getConnection',
            'end',
        ].forEach(method => {
            this.#pool[method] = (0, node_util_1.promisify)(this.#pool[method]);
        });
    }
    // impl Operator._query
    async _query(sql) {
        return await this.#pool.query(sql);
    }
    get pool() {
        return this.#pool;
    }
    get stats() {
        return {
            acquiringConnections: this.#pool._acquiringConnections.length,
            allConnections: this.#pool._allConnections.length,
            freeConnections: this.#pool._freeConnections.length,
            connectionQueue: this.#pool._connectionQueue.length,
        };
    }
    async getConnection() {
        try {
            const _conn = await this.#pool.getConnection();
            const conn = new connection_1.RDSConnection(_conn);
            if (this.beforeQueryHandlers.length > 0) {
                for (const handler of this.beforeQueryHandlers) {
                    conn.beforeQuery(handler);
                }
            }
            if (this.afterQueryHandlers.length > 0) {
                for (const handler of this.afterQueryHandlers) {
                    conn.afterQuery(handler);
                }
            }
            return conn;
        }
        catch (err) {
            if (err.name === 'Error') {
                err.name = 'RDSClientGetConnectionError';
            }
            throw err;
        }
    }
    /**
     * Begin a transaction
     *
     * @return {RDSTransaction} transaction instance
     */
    async beginTransaction() {
        const conn = await this.getConnection();
        try {
            await conn.beginTransaction();
        }
        catch (err) {
            conn.release();
            throw err;
        }
        const tran = new transaction_1.RDSTransaction(conn);
        if (this.beforeQueryHandlers.length > 0) {
            for (const handler of this.beforeQueryHandlers) {
                tran.beforeQuery(handler);
            }
        }
        if (this.afterQueryHandlers.length > 0) {
            for (const handler of this.afterQueryHandlers) {
                tran.afterQuery(handler);
            }
        }
        return tran;
    }
    /**
     * Auto commit or rollback on a transaction scope
     *
     * @param {Function} scope - scope with code
     * @param {Object} [ctx] - transaction env context, like koa's ctx.
     *   To make sure only one active transaction on this ctx.
     * @return {Object} - scope return result
     */
    async beginTransactionScope(scope, ctx) {
        ctx = ctx || {};
        if (!ctx._transactionConnection) {
            // Create only one conn if concurrent call `beginTransactionScope`
            ctx._transactionConnection = this.beginTransaction();
        }
        const tran = await ctx._transactionConnection;
        if (!ctx._transactionScopeCount) {
            ctx._transactionScopeCount = 1;
        }
        else {
            ctx._transactionScopeCount++;
        }
        try {
            const result = await scope(tran);
            ctx._transactionScopeCount--;
            if (ctx._transactionScopeCount === 0) {
                ctx._transactionConnection = null;
                await tran.commit();
            }
            return result;
        }
        catch (err) {
            if (ctx._transactionConnection) {
                ctx._transactionConnection = null;
                await tran.rollback();
            }
            throw err;
        }
    }
    /**
     * doomed to be rollbacked after transaction scope
     * useful on writing tests which are related with database
     *
     * @param {Function} scope - scope with code
     * @param {Object} [ctx] - transaction env context, like koa's ctx.
     *   To make sure only one active transaction on this ctx.
     * @return {Object} - scope return result
     */
    async beginDoomedTransactionScope(scope, ctx) {
        ctx = ctx || {};
        if (!ctx._transactionConnection) {
            ctx._transactionConnection = await this.beginTransaction();
            ctx._transactionScopeCount = 1;
        }
        else {
            ctx._transactionScopeCount++;
        }
        const tran = ctx._transactionConnection;
        try {
            const result = await scope(tran);
            ctx._transactionScopeCount--;
            if (ctx._transactionScopeCount === 0) {
                ctx._transactionConnection = null;
            }
            return result;
        }
        catch (err) {
            if (ctx._transactionConnection) {
                ctx._transactionConnection = null;
            }
            throw err;
        }
        finally {
            await tran.rollback();
        }
    }
    async end() {
        await this.#pool.end();
    }
}
exports.RDSClient = RDSClient;
