// egg-bin/dist/cmd/base.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCommand = void 0;
const node_util_1 = require("node:util");
const node_child_process_1 = require("node:child_process");
const artus_cli_1 = require("@artus-cli/artus-cli");
const debug = (0, node_util_1.debuglog)('egg-bin:base');
// only hook once and only when ever start any child.
const childs = new Set();
let hadHook = false;
function gracefull(proc) {
    // save child ref
    childs.add(proc);
    // only hook once
    /* c8 ignore else */
    if (!hadHook) {
        hadHook = true;
        let signal;
        ['SIGINT', 'SIGQUIT', 'SIGTERM'].forEach(event => {
            process.once(event, () => {
                signal = event;
                process.exit(0);
            });
        });
        process.once('exit', (code) => {
            for (const child of childs) {
                debug('process exit code: %o, kill child %o with %o', code, child.pid, signal);
                child.kill(signal);
            }
        });
    }
}
class ForkError extends Error {
    code;
    constructor(message, code) {
        super(message);
        this.code = code;
    }
}
let BaseCommand = class BaseCommand extends artus_cli_1.Command {
    dryRun;
    require;
    ctx;
    utils;
    // FIXME: should has a better way to init global args default value
    get base() {
        return this.ctx.args.base;
    }
    async run() {
        await this.utils.redirect(['--help']);
    }
    async formatRequires() {
        const requires = this.require ?? [];
        const eggRequire = this.ctx.args.pkgEgg.require;
        if (Array.isArray(eggRequire)) {
            for (const r of eggRequire) {
                requires.push(r);
            }
        }
        else if (typeof eggRequire === 'string' && eggRequire) {
            requires.push(eggRequire);
        }
        return requires;
    }
    async forkNode(modulePath, args, options = {}) {
        if (this.dryRun) {
            console.log('dry run: $ %o', `${process.execPath} ${modulePath} ${args.join(' ')}`);
            return;
        }
        const forkExecArgv = [
            ...this.ctx.args.execArgv || [],
            ...options.execArgv || [],
        ];
        options = {
            stdio: 'inherit',
            env: this.ctx.env,
            cwd: this.base,
            ...options,
            execArgv: forkExecArgv,
        };
        const proc = (0, node_child_process_1.fork)(modulePath, args, options);
        debug('Run fork pid: %o, `%s %s %s`', proc.pid, process.execPath, modulePath, args.join(' '));
        gracefull(proc);
        return new Promise((resolve, reject) => {
            proc.once('exit', code => {
                debug('fork pid: %o exit code %o', proc.pid, code);
                childs.delete(proc);
                if (code !== 0) {
                    const err = new ForkError(modulePath + ' ' + args.join(' ') + ' exit with code ' + code, code);
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
};
exports.BaseCommand = BaseCommand;
__decorate([
    (0, artus_cli_1.Option)({
        description: 'whether show full command script only, default is false',
        alias: 'd',
        type: 'boolean',
        default: false,
    }),
    __metadata("design:type", Boolean)
], BaseCommand.prototype, "dryRun", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'require the given module',
        alias: 'r',
        array: true,
        default: [],
    }),
    __metadata("design:type", Array)
], BaseCommand.prototype, "require", void 0);
__decorate([
    (0, artus_cli_1.Inject)(),
    __metadata("design:type", artus_cli_1.CommandContext)
], BaseCommand.prototype, "ctx", void 0);
__decorate([
    (0, artus_cli_1.Inject)(),
    __metadata("design:type", artus_cli_1.Utils)
], BaseCommand.prototype, "utils", void 0);
exports.BaseCommand = BaseCommand = __decorate([
    (0, artus_cli_1.DefineCommand)()
], BaseCommand);
