// egg-bin/dist/cmd/test.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestCommand = void 0;
const node_util_1 = require("node:util");
const node_os_1 = __importDefault(require("node:os"));
const promises_1 = __importDefault(require("node:fs/promises"));
const node_path_1 = __importDefault(require("node:path"));
const artus_cli_1 = require("@artus-cli/artus-cli");
const globby_1 = __importDefault(require("globby"));
const jest_changed_files_1 = require("jest-changed-files");
const base_1 = require("./base");
const debug = (0, node_util_1.debuglog)('egg-bin:test');
let TestCommand = class TestCommand extends base_1.BaseCommand {
    files;
    timeout;
    grep;
    changed;
    parallel;
    jobs;
    autoAgent;
    mochawesome;
    bail;
    async run() {
        try {
            await promises_1.default.access(this.base);
        }
        catch (err) {
            console.error('baseDir: %o not exists', this.base);
            throw err;
        }
        const mochaFile = process.env.MOCHA_FILE || require.resolve('mocha/bin/_mocha');
        if (this.parallel) {
            this.ctx.env.ENABLE_MOCHA_PARALLEL = 'true';
            if (this.autoAgent) {
                this.ctx.env.AUTO_AGENT = 'true';
            }
        }
        // set NODE_ENV=test, let egg application load unittest logic
        // https://eggjs.org/basics/env#difference-from-node_env
        this.ctx.env.NODE_ENV = 'test';
        debug('run test: %s %o', mochaFile, this.ctx.args);
        const mochaArgs = await this.formatMochaArgs();
        if (!mochaArgs)
            return;
        await this.forkNode(mochaFile, mochaArgs, {
            execArgv: [
                ...process.execArgv,
                // https://github.com/mochajs/mocha/issues/2640#issuecomment-1663388547
                '--unhandled-rejections=strict',
            ],
        });
    }
    async formatMochaArgs() {
        // collect require
        const requires = await this.formatRequires();
        try {
            const eggMockRegister = require.resolve('egg-mock/register', { paths: [this.base] });
            requires.push(eggMockRegister);
            debug('auto register egg-mock: %o', eggMockRegister);
        }
        catch (err) {
            // ignore egg-mock not exists
            debug('auto register egg-mock fail, can not require egg-mock on %o, error: %s', this.base, err.message);
        }
        // handle mochawesome enable
        let reporter = this.ctx.env.TEST_REPORTER;
        let reporterOptions = '';
        if (!reporter && this.mochawesome) {
            // use https://github.com/node-modules/mochawesome/pull/1 instead
            reporter = require.resolve('mochawesome-with-mocha');
            reporterOptions = 'reportDir=node_modules/.mochawesome-reports';
            if (this.parallel) {
                // https://github.com/adamgruber/mochawesome#parallel-mode
                requires.push(require.resolve('mochawesome-with-mocha/register'));
            }
        }
        const ext = this.ctx.args.typescript ? 'ts' : 'js';
        let pattern = this.files;
        // changed
        if (this.changed) {
            pattern = await this.getChangedTestFiles(this.base, ext);
            if (!pattern.length) {
                console.log('No changed test files');
                return;
            }
            debug('changed files: %o', pattern);
        }
        if (!pattern.length && process.env.TESTS) {
            pattern = process.env.TESTS.split(',');
        }
        // collect test files when nothing is changed
        if (!pattern.length) {
            pattern = [`test/**/*.test.${ext}`];
        }
        pattern = pattern.concat(['!test/fixtures', '!test/node_modules']);
        // expand glob and skip node_modules and fixtures
        const files = globby_1.default.sync(pattern, { cwd: this.base });
        files.sort();
        if (files.length === 0) {
            console.log(`No test files found with ${pattern}`);
            return;
        }
        // auto add setup file as the first test file
        const setupFile = node_path_1.default.join(this.base, `test/.setup.${ext}`);
        try {
            await promises_1.default.access(setupFile);
            files.unshift(setupFile);
        }
        catch {
            // ignore
        }
        return [
            this.dryRun ? '--dry-run' : '',
            // force exit
            '--exit',
            this.bail ? '--bail' : '',
            this.grep.map(pattern => `--grep='${pattern}'`).join(' '),
            this.timeout === false ? '--no-timeout' : `--timeout=${this.timeout}`,
            this.parallel ? '--parallel' : '',
            this.parallel && this.jobs ? `--jobs=${this.jobs}` : '',
            reporter ? `--reporter=${reporter}` : '',
            reporterOptions ? `--reporter-options=${reporterOptions}` : '',
            ...requires.map(r => `--require=${r}`),
            ...files,
        ].filter(a => a.trim());
    }
    async getChangedTestFiles(dir, ext) {
        const res = await (0, jest_changed_files_1.getChangedFilesForRoots)([node_path_1.default.join(dir, 'test')], {});
        const changedFiles = res.changedFiles;
        const files = [];
        for (let cf of changedFiles) {
            // only find test/**/*.test.(js|ts)
            if (cf.endsWith(`.test.${ext}`)) {
                // Patterns MUST use forward slashes (not backslashes)
                // This should be converted on Windows
                if (process.platform === 'win32') {
                    cf = cf.replace(/\\/g, '/');
                }
                files.push(cf);
            }
        }
        return files;
    }
};
exports.TestCommand = TestCommand;
__decorate([
    (0, artus_cli_1.Option)({
        default: [],
        array: true,
        type: 'string',
    }),
    __metadata("design:type", Array)
], TestCommand.prototype, "files", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'set test-case timeout in milliseconds, default is 60000',
        alias: 't',
        default: process.env.TEST_TIMEOUT ?? 60000,
    }),
    __metadata("design:type", Object)
], TestCommand.prototype, "timeout", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'only run tests matching <pattern>',
        alias: 'g',
        type: 'string',
        array: true,
        default: [],
    }),
    __metadata("design:type", Array)
], TestCommand.prototype, "grep", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'only test with changed files and match test/**/*.test.(js|ts), default is false',
        alias: 'c',
        type: 'boolean',
        default: false,
    }),
    __metadata("design:type", Boolean)
], TestCommand.prototype, "changed", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'mocha parallel mode, default is false',
        alias: 'p',
        type: 'boolean',
        default: false,
    }),
    __metadata("design:type", Boolean)
], TestCommand.prototype, "parallel", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'number of jobs to run in parallel',
        type: 'number',
        default: node_os_1.default.cpus().length - 1,
    }),
    __metadata("design:type", Number)
], TestCommand.prototype, "jobs", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'auto bootstrap agent in mocha master process, default is true',
        type: 'boolean',
        default: true,
    }),
    __metadata("design:type", Boolean)
], TestCommand.prototype, "autoAgent", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'enable mochawesome reporter, default is true',
        type: 'boolean',
        default: true,
    }),
    __metadata("design:type", Boolean)
], TestCommand.prototype, "mochawesome", void 0);
__decorate([
    (0, artus_cli_1.Option)({
        description: 'bbort ("bail") after first test failure',
        alias: 'b',
        type: 'boolean',
        default: false,
    }),
    __metadata("design:type", Boolean)
], TestCommand.prototype, "bail", void 0);
exports.TestCommand = TestCommand = __decorate([
    (0, artus_cli_1.DefineCommand)({
        command: 'test [files...]',
        description: 'Run the test',
        alias: ['t'],
    })
], TestCommand);
