// egg-ts-helper/dist/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.declMapping = exports.dtsComment = exports.dtsCommentRE = exports.eggInfoPath = exports.tmpDir = void 0;
const path_1 = __importDefault(require("path"));
const utils_1 = require("./utils");
const root = path_1.default.dirname(__dirname);
const packInfo = (0, utils_1.getPkgInfo)(root);
exports.tmpDir = path_1.default.join(root, '.tmp');
exports.eggInfoPath = path_1.default.resolve(exports.tmpDir, 'eggInfo.json');
exports.dtsCommentRE = new RegExp(`^\\/\\/ [\\w ]+ ${packInfo.name}(@\\d+\\.\\d+\\.\\d+)?`);
exports.dtsComment = `// This file is created by ${packInfo.name}@${packInfo.version}\n` +
    '// Do not modify this file!!!!!!!!!\n' +
    '/* eslint-disable */\n';
// mapping declaration in egg
exports.declMapping = {
    service: 'IService',
    controller: 'IController',
    ctx: 'Context',
    context: 'Context',
    app: 'Application',
    application: 'Application',
    agent: 'Agent',
    request: 'Request',
    response: 'Response',
    helper: 'IHelper',
    middleware: 'IMiddleware',
    config: 'EggAppConfig',
    plugin: 'EggPlugin',
};
