// @artus/injection/lib/container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constant_1 = require("./constant");
const types_1 = require("./types");
const util_1 = require("./util");
const error_1 = require("./error");
const lazy_helper_1 = require("./lazy_helper");
class Container {
    constructor(name) {
        this.name = name;
        this.registry = new Map();
        this.tags = new Map();
        this.handlerMap = new Map();
    }
    get(id, options = {}) {
        const md = this.getDefinition(id);
        if (!md) {
            if (options.noThrow) {
                return options.defaultValue;
            }
            throw new error_1.NotFoundError(id);
        }
        return this.getValue(md);
    }
    set(options) {
        var _a;
        if (options.id && !(0, util_1.isUndefined)(options.value)) {
            const md = Object.assign(Object.assign({}, options), { id: options.id, scope: (_a = options.scope) !== null && _a !== void 0 ? _a : types_1.ScopeEnum.SINGLETON });
            this.registry.set(md.id, md);
            /**
             * compatible with inject type identifier when identifier is string
             */
            if (md.type && (0, util_1.isClass)(md.type) && md.id !== md.type) {
                this.registry.set(md.type, md);
            }
            return this;
        }
        const { type, id, scope, scopeEscape } = this.getDefinedMetaData(options);
        const md = Object.assign(Object.assign({}, options), { id,
            type,
            scope,
            scopeEscape });
        if (type) {
            const args = (0, util_1.getMetadata)(constant_1.CLASS_CONSTRUCTOR_ARGS, type);
            const props = (0, util_1.recursiveGetMetadata)(constant_1.CLASS_PROPERTY, type);
            const handlerArgs = (0, util_1.getMetadata)(constant_1.INJECT_HANDLER_ARGS, type);
            const handlerProps = (0, util_1.recursiveGetMetadata)(constant_1.INJECT_HANDLER_PROPS, type);
            md.constructorArgs = (args !== null && args !== void 0 ? args : []).concat(handlerArgs !== null && handlerArgs !== void 0 ? handlerArgs : []);
            md.properties = (props !== null && props !== void 0 ? props : []).concat(handlerProps !== null && handlerProps !== void 0 ? handlerProps : []);
            /**
             * compatible with inject type identifier when identifier is string
             */
            if (md.id !== type) {
                this.registry.set(type, md);
            }
            this.handleTag(type);
        }
        this.registry.set(md.id, md);
        if (md.eager && md.scope !== types_1.ScopeEnum.TRANSIENT) {
            this.get(md.id);
        }
        return this;
    }
    getDefinition(id) {
        return this.registry.get(id);
    }
    getInjectableByTag(tag) {
        const result = this.tags.get(tag);
        return result ? [...result] : [];
    }
    getByTag(tag) {
        const clazzes = this.getInjectableByTag(tag);
        return clazzes.map(clazz => this.get(clazz));
    }
    registerHandler(name, handler) {
        this.handlerMap.set(name, handler);
    }
    getHandler(name) {
        return this.handlerMap.get(name);
    }
    hasValue(options) {
        const { id } = this.getDefinedMetaData(options);
        const md = this.getDefinition(id);
        return !!md && !(0, util_1.isUndefined)(md.value);
    }
    getValueByMetadata(md) {
        if (md.handler) {
            return this.resolveHandler(md.handler, md.id);
        }
        return this.get(md.id, {
            noThrow: md.noThrow,
            defaultValue: md.defaultValue,
        });
    }
    getValue(md) {
        if (!(0, util_1.isUndefined)(md.value)) {
            return md.value;
        }
        let value;
        if (md.factory) {
            value = md.factory(md.id, this);
        }
        if (!value && md.type) {
            const clazz = md.type;
            const params = this.resolveParams(clazz, md);
            value = new clazz(...params);
            this.resolveProps(value, md);
        }
        if (md.scope === types_1.ScopeEnum.SINGLETON) {
            md.value = value;
        }
        return value;
    }
    getDefinedMetaData(options) {
        var _a, _b, _c, _d;
        let { type, id, scope = types_1.ScopeEnum.SINGLETON, factory, scopeEscape } = options;
        if (!type) {
            if (id && (0, util_1.isClass)(id)) {
                type = id;
            }
        }
        if (!type && !factory) {
            throw new error_1.NoTypeError(`injectable ${id === null || id === void 0 ? void 0 : id.toString()}`);
        }
        if (factory && !(0, util_1.isFunction)(factory)) {
            throw new error_1.InjectionError('factory option must be function');
        }
        if (type) {
            const targetMd = (0, util_1.getMetadata)(constant_1.CLASS_CONSTRUCTOR, type) || {};
            id = (_b = (_a = targetMd.id) !== null && _a !== void 0 ? _a : id) !== null && _b !== void 0 ? _b : type;
            scope = (_c = targetMd.scope) !== null && _c !== void 0 ? _c : scope;
            scopeEscape = (_d = targetMd.scopeEscape) !== null && _d !== void 0 ? _d : scopeEscape;
        }
        if (!id && factory) {
            throw new error_1.NoIdentifierError(`injectable with factory option`);
        }
        return { type, id: id, scope, scopeEscape };
    }
    resolveParams(clazz, md) {
        var _a;
        let { constructorArgs: args = [] } = md;
        const params = [];
        if (!args || !args.length) {
            args = ((_a = (0, util_1.getParamMetadata)(clazz)) !== null && _a !== void 0 ? _a : []).map((ele, index) => ({
                id: ele,
                index,
            }));
        }
        args.forEach(arg => {
            if ((0, util_1.isPrimitiveFunction)(arg.id)) {
                return;
            }
            if (!arg.handler) {
                this.checkScope(md, arg.id, arg.index);
            }
            params[arg.index] = this.getValueByMetadata(arg);
        });
        return params;
    }
    resolveProps(instance, md) {
        const { properties = [] } = md;
        properties.forEach(prop => {
            if (!prop.handler) {
                this.checkScope(md, prop.id, prop.propertyName);
            }
            if (prop.lazy) {
                return (0, lazy_helper_1.createLazyProperty)(instance, prop, this);
            }
            instance[prop.propertyName] = this.getValueByMetadata(prop);
        });
    }
    handleTag(target) {
        let tags = Reflect.getOwnMetadata(constant_1.CLASS_TAG, target);
        if (!tags) {
            return;
        }
        if (!Array.isArray(tags)) {
            tags = [tags];
        }
        tags.forEach(tag => {
            if (!this.tags.has(tag)) {
                this.tags.set(tag, new Set());
            }
            this.tags.get(tag).add(target);
        });
    }
    resolveHandler(handlerName, id) {
        const handler = this.getHandler(handlerName);
        if (!handler) {
            throw new error_1.NoHandlerError(handlerName.toString());
        }
        return id ? handler(id, this) : handler(this);
    }
    /**
     * check rule
     * The first column is the class scope and the first row is the property scope
     * ----------------------------------------
     *          |singleton|execution |transient
     * ----------------------------------------
     * singleton|✅        |❌        |✅
     * ----------------------------------------
     * execution|✅        |✅        |✅
     * ----------------------------------------
     * transient|✅        |❓        |✅
     * ----------------------------------------
     */
    checkScope(metadata, id, propertyOrIndex) {
        const { scope } = metadata;
        if (scope === types_1.ScopeEnum.EXECUTION || scope === types_1.ScopeEnum.TRANSIENT) {
            return;
        }
        const propMetadata = this.getDefinition(id);
        if ((propMetadata === null || propMetadata === void 0 ? void 0 : propMetadata.scope) === types_1.ScopeEnum.EXECUTION && !propMetadata.scopeEscape) {
            throw new error_1.ScopeEscapeError(metadata.type, propertyOrIndex, scope, propMetadata.scope);
        }
    }
}
exports.default = Container;
