// @artus/injection/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopeEscapeError = exports.LazyInjectConstructorError = exports.InjectionError = exports.NoIdentifierError = exports.NoHandlerError = exports.NotFoundError = exports.NoTypeError = exports.CannotInjectValueError = void 0;
const base_error_1 = require("./base_error");
class CannotInjectValueError extends (0, base_error_1.createErrorClass)('CannotInjectValueError') {
    constructor(target, propertyOrIndex) {
        super(() => {
            let message = `'${target.name}.${String(propertyOrIndex)}'`;
            if (typeof propertyOrIndex === 'number') {
                message = `'${target.name}' constructor argument at index '${propertyOrIndex}'`;
            }
            return `[@artus/injection] Cannot inject value into ${message}, maybe inject identifier is undefined or type is primitive type`;
        });
    }
}
exports.CannotInjectValueError = CannotInjectValueError;
class NoTypeError extends (0, base_error_1.createErrorClass)('NoTypeError') {
    constructor(message) {
        super(`[@artus/injection] type is required: ${message}`);
    }
}
exports.NoTypeError = NoTypeError;
class NotFoundError extends (0, base_error_1.createErrorClass)('NotFoundError') {
    constructor(identifier) {
        const normalizedIdentifier = typeof identifier === 'function' ? identifier.name : (identifier !== null && identifier !== void 0 ? identifier : 'Unknown').toString();
        super(() => {
            return (`[@artus/injection] with "${normalizedIdentifier}" ` +
                `identifier was not found in the container. `);
        });
    }
}
exports.NotFoundError = NotFoundError;
class NoHandlerError extends (0, base_error_1.createErrorClass)('NoHandlerError') {
    constructor(handler) {
        super(() => {
            return `[@artus/injection] "${handler}" handler was not found in the container.`;
        });
    }
}
exports.NoHandlerError = NoHandlerError;
class NoIdentifierError extends (0, base_error_1.createErrorClass)('NoIdentifierError') {
    constructor(message) {
        super(`[@artus/injection] id is required: ${message}`);
    }
}
exports.NoIdentifierError = NoIdentifierError;
class InjectionError extends (0, base_error_1.createErrorClass)('InjectionError') {
    constructor(message) {
        super(`[@artus/injection] ${message}`);
    }
}
exports.InjectionError = InjectionError;
class LazyInjectConstructorError extends (0, base_error_1.createErrorClass)('LazyInjectConstructor') {
    constructor(name) {
        super(`[@artus/injection] cannot inject '${name}' constructor argument by lazy`);
    }
}
exports.LazyInjectConstructorError = LazyInjectConstructorError;
class ScopeEscapeError extends (0, base_error_1.createErrorClass)('ScopeEscapeError') {
    constructor(target, propertyOrIndex, classScope, propScope) {
        super(() => {
            let message = `property '${String(propertyOrIndex)}'`;
            if (typeof propertyOrIndex === 'number') {
                message = `constructor argument at index '${propertyOrIndex}'`;
            }
            return `[@artus/injection] '${target.name}' with '${classScope}' scope cannot be injected ${message} with '${propScope}' scope`;
        });
    }
}
exports.ScopeEscapeError = ScopeEscapeError;
