// @artus/injection/lib/execution_container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const constant_1 = require("./constant");
const container_1 = tslib_1.__importDefault(require("./container"));
const types_1 = require("./types");
const error_1 = require("./error");
const util_1 = require("./util");
class ExecutionContainer extends container_1.default {
    constructor(ctx, parent) {
        super('execution');
        this.parent = parent;
        this.ctx = ctx;
        this.set({ id: constant_1.EXECUTION_CONTEXT_KEY, value: ctx });
    }
    get(id, options = {}) {
        const md = this.getDefinition(id);
        if (!md) {
            if (options.noThrow) {
                return options.defaultValue;
            }
            throw new error_1.NotFoundError(id);
        }
        const value = this.getValue(md);
        if (md.scope === types_1.ScopeEnum.EXECUTION) {
            this.setValue(md, value);
        }
        return value;
    }
    getDefinition(id) {
        var _a;
        return (_a = super.getDefinition(id)) !== null && _a !== void 0 ? _a : this.parent.getDefinition(id);
    }
    getInjectableByTag(tag) {
        let tags = super.getInjectableByTag(tag);
        if (!tags || tags.length === 0) {
            tags = this.parent.getInjectableByTag(tag);
        }
        return tags;
    }
    getCtx() {
        return this.ctx;
    }
    getHandler(name) {
        var _a;
        return (_a = super.getHandler(name)) !== null && _a !== void 0 ? _a : this.parent.getHandler(name);
    }
    setValue(md, value) {
        if (!(0, util_1.isUndefined)(md.value)) {
            return;
        }
        if (md.type && md.id !== md.type) {
            this.set(Object.assign(Object.assign({}, md), { id: md.type, value }));
        }
        this.set(Object.assign(Object.assign({}, md), { id: md.id, value }));
    }
}
exports.default = ExecutionContainer;
