// @artus/injection/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPrimitiveFunction = exports.isFunction = exports.isObject = exports.isUndefined = exports.isNumber = exports.isClass = exports.isInjectable = exports.addTag = exports.getDesignTypeMetadata = exports.getParamMetadata = exports.recursiveGetMetadata = exports.setMetadata = exports.getMetadata = void 0;
const constant_1 = require("./constant");
const functionPrototype = Object.getPrototypeOf(Function);
function getMetadata(metadataKey, target, propertyKey) {
    if (propertyKey) {
        return Reflect.getOwnMetadata(metadataKey, target, propertyKey);
    }
    return Reflect.getOwnMetadata(metadataKey, target);
}
exports.getMetadata = getMetadata;
function setMetadata(metadataKey, value, target, propertyKey) {
    if (propertyKey) {
        Reflect.defineMetadata(metadataKey, value, target, propertyKey);
    }
    else {
        Reflect.defineMetadata(metadataKey, value, target);
    }
}
exports.setMetadata = setMetadata;
/**
 * recursive get class and super class metadata
 * @param metadataKey
 * @param target
 * @param propertyKey
 * @returns
 */
function recursiveGetMetadata(metadataKey, target, propertyKey) {
    let metadatas = [];
    const metadata = getMetadata(metadataKey, target, propertyKey);
    if (metadata) {
        metadatas = metadatas.concat(metadata);
    }
    let proto = Object.getPrototypeOf(target);
    while (proto !== null && proto !== functionPrototype) {
        const metadata = getMetadata(metadataKey, proto, propertyKey);
        if (metadata) {
            metadatas = metadatas.concat(metadata);
        }
        proto = Object.getPrototypeOf(proto);
    }
    return metadatas;
}
exports.recursiveGetMetadata = recursiveGetMetadata;
/**
 * get constructor parameter types
 * @param clazz
 * @returns
 */
function getParamMetadata(clazz) {
    return Reflect.getMetadata('design:paramtypes', clazz);
}
exports.getParamMetadata = getParamMetadata;
/**
 * get the property type
 * @param clazz
 * @param property
 * @returns
 */
function getDesignTypeMetadata(clazz, property) {
    return Reflect.getMetadata('design:type', clazz, property);
}
exports.getDesignTypeMetadata = getDesignTypeMetadata;
function addTag(tag, target) {
    let tags = Reflect.getOwnMetadata(constant_1.CLASS_TAG, target);
    if (!tags) {
        tags = [];
        Reflect.defineMetadata(constant_1.CLASS_TAG, tags, target);
    }
    tags.push(tag);
}
exports.addTag = addTag;
function isInjectable(target) {
    return Reflect.hasOwnMetadata(constant_1.CLASS_CONSTRUCTOR, target);
}
exports.isInjectable = isInjectable;
function isClass(clazz) {
    if (typeof clazz !== 'function') {
        return false;
    }
    const fnStr = Function.prototype.toString.call(clazz);
    return (fnStr.substring(0, 5) === 'class' ||
        Boolean(~fnStr.indexOf('classCallCheck(')) ||
        Boolean(~fnStr.indexOf('TypeError("Cannot call a class as a function")')));
}
exports.isClass = isClass;
function isNumber(value) {
    return typeof value === 'number';
}
exports.isNumber = isNumber;
function isUndefined(value) {
    return typeof value === 'undefined';
}
exports.isUndefined = isUndefined;
function isObject(value) {
    return typeof value === 'object';
}
exports.isObject = isObject;
function isFunction(value) {
    return typeof value === 'function';
}
exports.isFunction = isFunction;
function isPrimitiveFunction(value) {
    return ['String', 'Boolean', 'Number', 'Object'].includes(value.name);
}
exports.isPrimitiveFunction = isPrimitiveFunction;
